executor/hash_table.go

```diff
@@ -149,14 +172,51 @@ func (c *hashRowContainer) matchJoinKey(buildRow, probeRow chunk.Row, probeHCtx
 		probeRow, probeHCtx.allTypes, probeHCtx.keyColIdx)
 }
 
+func (c *hashRowContainer) spillToDisk() (err error) {
+	N := c.records.NumChunks()
+	c.recordsInDisk = chunk.NewListInDisk(c.hCtx.allTypes)
+	for i := 0; i < N; i++ {
+		chk := c.records.GetChunk(i)
+		err = c.recordsInDisk.Add(chk)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// alreadySpilled indicates that records have spilled out into disk.
+func (c *hashRowContainer) alreadySpilled() bool { return c.recordsInDisk != nil }
+
+// alreadySpilledSafe indicates that records have spilled out into disk. It's thread-safe.
+func (c *hashRowContainer) alreadySpilledSafe() bool { return atomic.LoadUint32(&c.spilled) == 1 }
+
 // PutChunk puts a chunk into hashRowContainer and build hash map. It's not thread-safe.
 // key of hash table: hash value of key columns
 // value of hash table: RowPtr of the corresponded row
 func (c *hashRowContainer) PutChunk(chk *chunk.Chunk) error {
-	chkIdx := uint32(c.records.NumChunks())
+	var chkIdx uint32
+	if c.alreadySpilled() {
+		// append chk to disk.
+		chkIdx = uint32(c.recordsInDisk.NumChunks())
+		err := c.recordsInDisk.Add(chk)
+		if err != nil {
+			return err
+		}
+	} else {
+		chkIdx = uint32(c.records.NumChunks())
+		c.records.Add(chk)
+		if c.exceeded {
```

Comment:
can we put this into `spillDiskAction.Action()`?

Reply:
Hmm, it’s a easy implementation if we keep Putting chunk into memory and into disk in the same thread. Otherwise, we must keep syncing the state of whether the data has been spilled within two goroutines.

Reply:
Could you add a comment on `spillDiskAction.Action()` to explain that the action is triggered outside the scope? So that the readers may not confused.